dash/src/Man/man1/dummyasciiproducer.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.35.
.TH DUMMYASCIIPRODUCER "1" "February 2006" "dummyasciiproducer 1.82" "User Commands"
.SH NAME
dummyasciiproducer \- manual page for dummyasciiproducer 1.82
.SH SYNOPSIS
.B dummyasciiproducer
[\fIOPTION\fR]...
.SH DESCRIPTION
Output random sized records from a dummy ASCII records for consumption
by a DAQ ASCII on\-ramp.  This program produces printable ASCII
records separated by a separator character.  This program is intended
as an ASCII on\-ramp test program.
.SH OPTIONS
.TP
\fB\-\-version\fR
Output version information and exit
.TP
\fB\-\-help\fR
Display this help and exit
.TP
\fB\-\-debug\fR=\fISECONDS\fR
Output a debugging command and wait SECONDS
before beginning execution of main loop
.TP
\fB\-\-size\fR=\fIINTEGER\fR
Specify a fixed record size.  The default
is to generate random sized records up to a
maximum of 64K.
.TP
\fB\-\-iterations\fR=\fIINTEGER\fR Specify the number of data records this
producer should emit prior to exiting.
.HP
\fB\-\-delay\fR=\fIMICROSECONDS\fR Output occurs at a delay rate of MICROSECONDS
.TP
\fB\-\-separator\fR=\fICHARCTER\fR Specify that CHARCTER should be used as the
separator character.  Default is a new line
character.
.SH EXAMPLES
.IP
dummyproducer \fB\-\-run\fR=\fI45\fR | asciionramp \fB\-\-packet\fR=\fI104\fR dummyasciiproducer |
.IP
segmenter /evt/data/output
.IP
Inject output from dummybinaryoneramp into the data stream.
Records produced from data from dummyasciiproducer will have a type
of 104.  If dummyasciiproducer terminates, it will not be restarted.
.IP
dummyproducer \fB\-\-run\fR=\fI46\fR | asciionramp \fB\-\-packet\fR=\fI103\fR \fB\-\-restart\fR
.IP
dummyasciiproducer \fB\-\-iterations\fR=\fI10\fR \fB\-\-separator=\fR'\e\ef' |
segmenter /evt/data/output
.IP
Inject output from dummyasciiproducer into the data stream.
Records produced from data produced by dummyasciiproducer will
have a type of 103.  Dummybinaryproducer will emit 10 records each
time it is executed and then exit.  A formfeed character will
be used as the separator character.  When dummyasciiproducer
terminates, it is restarted by asciionramp.
.SH AUTHOR
Written by E. Kasten.
.SH COPYRIGHT
Copyright \(co 2005 by the Board of Trustees of Michigan State University.
You may use this software under the terms of the GNU public license
(GPL).  The terms of this license are described at:
http://www.gnu.org/licenses/gpl.txt


dash/src/Pipeline/Sources/dummyasciiproducer.cc
/*=========================================================================*\
| Copyright (C) 2005 by the Board of Trustees of Michigan State University. |
| You may use this software under the terms of the GNU public license       |
| (GPL).  The terms of this license are described at:                       |
| http://www.gnu.org/licenses/gpl.txt                                       |
|                                                                           |
| Written by: E. Kasten                                                     |
\*=========================================================================*/

using namespace std;

#include <stdlib.h>
#include <stdio.h>
#include <iostream>
#include <string>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <netinet/in.h>

#ifndef DAQHWYAPI_H
#include <dshapi/daqhwyapi.h>
#endif

#ifndef DUMMYASCIIPRODUCER_VERSION_H
#include <dshusage/dummyasciiproducer_version.h>
#endif

#ifndef DUMMYASCIIPRODUCER_HELP_H
#include <dshusage/dummyasciiproducer_help.h>
#endif

using namespace daqhwyapi;

/**
* @var dummyasciiproducer_rtexception
* @brief Exception to throw for general runtime errors.
*
* Exception to throw for for general runtime errors.
*/
static RuntimeException dummyasciiproducer_rtexception;

#define MAX_RECORD_SIZE 64536

static char *fillarray = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
static int fillarraylen = strlen(fillarray);

/*==============================================================*/
/**
* @class DummyAsciiProducer
* @brief Produces dummy DAQ records for testing.
*
* Produces dummy DAQ records for testing pipeline elements.
*
* @author  Eric Kasten
* @version 1.0.0
*/
class DummyAsciiProducer: public Main {
  private: unsigned long my_debug_delay;
  private: unsigned long my_delay;
  private: unsigned long my_record_size;
  private: int my_iter_count;
  private: char my_sepchar;

  /*==============================================================*/
  /** @fn DummyAsciiProducer()
  * @brief Default constructor.
  *
  * Default constructor.
  *
  * @param None
  * @return this
  */
  public: DummyAsciiProducer() { 
    my_debug_delay = 0;
    my_delay = 0;
    my_record_size = 0;  // Random sized buffers
    my_iter_count = -1; // Run forever
    my_sepchar = '\n';
  }

  /*==============================================================*/
  /** @fn ~DummyAsciiProducer()
  * @brief Destructor
  *
  * Destructor.
  *
  * @param None
  * @return None
  */
  public: virtual ~DummyAsciiProducer() { }

  /*==============================================================*/
  /** @fn void version()
  * @brief Display program version.
  *
  * Display program version information.
  *
  * @param None
  * @return None
  */
  private: void version() {
    cerr << version_string << endl;
  }

  /*==============================================================*/
  /** @fn void usage()
  * @brief Display program usage.
  *
  * Display program usage.
  *
  * @param None
  * @return None
  */
  private: void usage() {
    cerr << help_string << endl;
  }

  /*==============================================================*/
  /** @fn bool parseCommandline(int argc,char *argv[])
  * @brief Parse the program command line.
  *
  * Parse the program command line and set parameters
  * as required.
  *
  * @param argc Number of commandline arguments including command name.
  * @param argv Array of commandline arguments.
  * @return If command line parsing was successful.
  */
  private: bool parseCommandline(int argc,char *argv[]) {
    // Parse the command line arguments
    int i = 1;  // Skip command name
    while (i < argc) { 
      String argstr;
      String parmstr;

      // Split arguments with options.
      if (argv[i] != NULL) {
        char *p = strchr(argv[i],'='); 
        if (p != NULL) {
          (*p) = '\0'; p++;
          argstr = argv[i];
          parmstr = p;
          (*(--p)) = '=';
        } else {
          argstr = argv[i];
        }
      }

      // Parse argment strings
      if (argstr[0] == '-') { // Option argument
        if (argstr.equals("--version")) { 
          version();
          exit(0);
        } else if (argstr.equals("--help")) { 
          usage();
          exit(0);
        } else if (argstr.equals("--debug")) { 
          if (parmstr.size() > 0) {
            my_debug_delay = StringUtils::parseInteger(parmstr);
          } else {
            cerr << "Argument \"--debug\" requires an integer parameter." << endl;
            return false;
          }
        } else if (argstr.equals("--size")) { 
          if (parmstr.size() > 0) {
            my_record_size = StringUtils::parseInteger(parmstr);
          } else {
            cerr << "Argument \"--size\" requires an integer parameter." << endl;
            return false;
          }
        } else if (argstr.equals("--separator")) { 
          if (parmstr.size() > 0) {
            String str = parmstr;
            DSHUtils::convertEscapeCharacters(str);
            my_sepchar = str[0]; 
          } else {
            cerr << "Argument \"--separator\" requires an character parameter." << endl;
            return false;
          }
        } else if (argstr.equals("--iterations")) { 
          if (parmstr.size() > 0) {
            my_iter_count = StringUtils::parseInteger(parmstr);
          } else {
            cerr << "Argument \"--iterations\" requires an integer parameter." << endl;
            return false;
          }
        } else if (argstr.equals("--delay")) { 
          if (parmstr.size() > 0) {
            my_delay = StringUtils::parseInteger(parmstr);
          } else {
            cerr << "Argument \"--delay\" requires an integer parameter." << endl;
            return false;
          }
        } else { // Unknown paramter
          cerr << "Argument \"" << (argstr.c_str()) << "\" is unrecognized." << endl;
          return false; 
        }
      } 
      i++;
    }

    return true;
  }

  /*==============================================================*/
  /** @fn void mainloop()
  * @brief Run the main loop.
  *
  * Run the main loop. Create records and write them to
  * stdout.
  * 
  * @param None
  * @return None
  */
  private: void mainloop() {
    // Create a data buffer
    ByteArray data;

    // Seed random
    ::srandom(System.currentTimeMillis());

    // And begin writing data
    int fcnt = 0;
    for (;;) {
      // Wait a bit
      if (my_delay > 0) usleep(my_delay);

      // Write length and data
      int recsiz = my_record_size;
      if (recsiz <= 0) recsiz = (::random() % MAX_RECORD_SIZE);
      data.renew(recsiz); 
      memset(data.elements,fillarray[fcnt],recsiz);
      ::write(1,(void*)(data.elements),recsiz); // Data 
      ::write(1,(void*)(&my_sepchar),1);  // Separator

      fcnt++;
      if (fcnt >= fillarraylen) fcnt = 0;

      // If we are iterating a fixed number of times...
      if (my_iter_count > 0) {
        my_iter_count--;
        if (my_iter_count <= 0) break;  // We are done
      }
    }
  }

  /*==============================================================*/
  /** @fn void main(int argc,char *argv[]) 
  * @brief Mainline startup method.
  *
  * Mainline startup method.
  *
  * @param argc Number of commandline arguments including command name.
  * @param argv Array of commandline arguments.
  * @return None  
  */
  public: void main(int argc,char *argv[]) {
    if (!parseCommandline(argc,argv)) {
      usage();
      exit(-1);
    }

    // Do we want to attach a debugger?
    if (my_debug_delay > 0) {
      System.attachDebugger(argv[0],my_debug_delay);
    }
  
    // Everything appears ok.
    mainloop();
  }
};

DummyAsciiProducer dummyasciiproducer;
